test/error_messages.coffee

```diff
@@ -1925,3 +1925,29 @@ test "`new.target` is only allowed meta property", ->
     -> new.something
        ^^^^^^^^^^^^^
   '''
+
+test "#4834: dynamic import requires exactly one argument", ->
+  assertErrorFormat '''
+    import()
+  ''', '''
+    [stdin]:1:1: error: import() requires exactly one argument
+    import()
+    ^^^^^^^^
+  '''
+
+  assertErrorFormat '''
+    import('x', {})
+  ''', '''
+    [stdin]:1:1: error: import() requires exactly one argument
+    import('x', {})
+    ^^^^^^^^^^^^^^^
+  '''
+
+test "#4834: dynamic import requires explicit call parentheses", ->
+  assertErrorFormat '''
+    promise = import 'foo'
+  ''', '''
+    [stdin]:1:23: error: unexpected end of input
```

Comment:
This isn’t a deal-breaker, but I suppose we can’t get a better error message here? Probably not, I’m assuming, because this is a grammar fail?

Reply:
Ya I think we're at the mercy of the grammar there

Reply:
Maybe a fake import rule which throws an error?
We don’t do that too often, but...

Reply:
@vendethiel can you point to an example? I think it might be worth it in this case because the required parentheses will likely be surprising.

Reply:
@GeoffreyBooth @vendethiel I agree it'd be nice. But I don't see an obvious way to implement it

Part of why I went with required parentheses is that @jashkenas' [suggestion](https://github.com/jashkenas/coffeescript/issues/4834#issuecomment-465714019) seemed tricky to implement (though I additionally also agree with the apparent consensus that required parentheses, though perhaps unprecedented, are actually a nice disambiguator readability-wise)

Recall that `import 'a'` is valid static import syntax. You couldn't add implicit call parens since at rewriter time you have no idea if you're in statement or "nested expression" context, so the token sequence (visible to the grammar) for dynamic `import 'a'` and static `import 'a'` would be identical

So I don't think you could have a separate non-conflicting grammar rule for the *expression* `import 'a'` (since `Expression`s and `Statement`s are both `Line`s)

So I think that implies that you'd have to parse dynamic `import 'a'` and static `import 'a'` with the same grammar rule (ie instantiate the same node class, which would have to be a hack that represented both dynamic imports and static imports)...and then only at compile time would you know reliably (based on `o.level`) if you were a "nested expression" and therefore a dynamic import (whether for purposes of allowing implicit-call-paren dynamic imports or just throwing a nicer error message)?

If I'm correct, I think that's ugly enough that it wouldn't be worth doing period, let alone just for the sake of a nicer error message

And I can't think of any other "heuristic" way of catching many common cases - I guess in theory you could add a rewriter rule to look for eg the token sequence `AWAIT` `IMPORT` [something other than `(`] and throw a custom error message then? But even that doesn't necessarily seem worth it imo?

Reply:
If there's no easy way to do it, it's not worth doing. I was thinking a new grammar rule that just threw an error might do it, then that's a one line addition. The error isn't with new node classes and so on.

Reply:
> ...and then only at compile time would you know reliably (based on o.level)

`ModuleDeclaration`'s `checkScope` already has error checking depending on `o.indent.length` (which also relies on `--bare` implied for modules, but that's fine). 

> If I'm correct, I think that's ugly enough that it wouldn't be worth doing period, let alone just for the sake of a nicer error message

At the same time, the current impl. impacts lexer, grammar and nodes. -- but yes, it's better than moving this down somewhere else.

> So I don't think you could have a separate non-conflicting grammar rule for the _expression_ `import 'a'` (since `Expression`s and `Statement`s are both `Line`s)

I was thinking to check for a `TERMINATOR` before the `(`-less `IMPORT`, but this breaks cases such as:

```coffee
f(1; import("a"))
# or the long form:
f 1,
  import("a")
```

Reply:
@vendethiel So do you think there’s a not-too-invasive way to add the error?

We can also just merge this and add a better error as a follow-on PR.

Reply:
I'm not totally clear on what case @vendethiel is suggesting we could catch - again, since "`(`-less `IMPORT`" is indistinguishable from the token sequence of a static import as I understand it, you couldn't just check for a preceding `TERMINATOR` since that would be a valid token sequence for static imports

Reply:
> you couldn't just check for a preceding `TERMINATOR` since that would be a valid token sequence for static imports

If it was only a valid token for static imports, it'd be easy-ish to generate a correct error message in some cases, but that's not the case.
The one feature usable is jison' start conditions, but we don't do that at all, and it's probably better not to add context for such a thing.

Reply:
Okay so @vendethiel @helixbass I’m assuming you both agree that we should merge this as is?

Reply:
Yes. I see no simple way to make the error message more palatable right now.

Reply:
👍